### PR TITLE
Support 512K "RCT6" in BTT SKR Mini

### PR DESCRIPTION
**Pull request by thisiskeithb**
### Description

BigTreeTech's SKR Mini line of boards with `STM32F103RCT6` MCUs can use 512k of flash (and potentially 64k of ram, but is untested/implemented in this PR) as if they were fitted with an `STM32F103RET6`.

This is similar to the double flash "feature" of `STM32F103C8T6`/bluepill boards.

USB is also now enabled by default since we don't have to worry about space.

*Thanks to @AletheianAlex via [this video](https://www.youtube.com/watch?v=EBIahC1P2e0) & [addendum](https://www.youtube.com/watch?v=7Utygr71p8s), @steenerson via [this reddit thread](https://old.reddit.com/r/ender3/comments/dsmf6x/512k_flash_working_on_skr_mini_e3_v12/) , and  reddit user yuchen_ting for[ initially posting about this](https://old.reddit.com/r/ender3/comments/dsct5x/in_fact_the_flash_size_of_skr_mini_e3_v12_is_512kb/).

### Benefits

image

### Related Issues

#15254, #15398

### CODE REVIEW

**Comment by damvcoool** [comment]
the SKR Mini E3 v1.2 actually has 256k, since it uses STM32F103RCT6, according to the STMicroelectronics product site.

I believe some SKR Mini E3 were shipped with STM32F103RE variant which does have 512k, but that's the exception not the rule.

Edit: forgot to add the link to STMicroelectronics product site.
https://www.st.com/en/microcontrollers-microprocessors/stm32f103.html#products

**Comment by TheZoker** [comment]
@damvcoool Yes but the point is, that version 1.2 with the F103RC still has 512kb, have a look at this Reddit thread where the guys confirmed this:
https://www.reddit.com/r/ender3/comments/dsmf6x/512k_flash_working_on_skr_mini_e3_v12/

**Comment by sjasonsmith** [comment]
> @damvcoool Yes but the point is, that version 1.2 with the F103RC still has 512kb, have a look at this Reddit thread where the guys confirmed this

It seems risky to use more than the documented amount of flash by default. It is the equivalent of overclocking a CPU. It might work for some chips, but isn't guaranteed to work for all.

For example, they may allow failures in half their flash and bin those parts to the RC part number, even though the chip is otherwise the same as an RE. Some users might have a perfectly reliable 512k, while other users may have a reliable 256k and a flaky 256k.

They might even design their chip to allow either half of the flash to fail testing, and they choose which half to enable when they package the die. Depending on which half is enabled you may or may not even have the other half connected.

Maybe none of this is happening, but it is a risk we take by using these chips off-spec.

**Comment by sjasonsmith** [comment]
I do support giving people the option to build with 512k if they choose, without making it the default for RC chips. Both my mini boards have RE chips on them, so it would be nice to easily make use of the space, or allow users with RC chips to attempt it _if they choose_.

**Comment by TheZoker** [comment]
@sjasonsmith I totally agree

**Comment by steenerson** [comment]
Thanks for prettying this up!

This has been done with bluepill F103C8 boards [going back years](https://github.com/texane/stlink/issues/516), and I spent 20 minutes searching last week and couldn't find a single example of anyone posting that they didn't have the doubled space available or that it didn't work correctly. I personally think ST would have fully masked these to 256K chips and not just left it available but undocumented if they weren't sure the full 512K was working, there's just too much downside to shipping unverified hardware. They even report the full size if you connect to the USART bootloader instead of st-link.

That being said, I tend to agree that 256K vs 512K should still be left as an option. I was thinking about making 256K default but adding a compile error for >224K program size that suggests trying 512K and making the change to 512K be a single config option somewhere. But haven't gotten anywhere with it.

**Comment by thisiskeithb** [comment]
> This has been done with bluepill F103C8 boards going back years,

I'm usually one for being cautious with things affecting lots users, but this is why I made it default. We can always make it an option if it becomes an issue.

**Comment by swilkens** [comment]
> Enable SKR Mini USB by default 

It is wise to switch to this if the OEM manuals assume we default to non-usb? (i.e. the last section of this https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/tree/master/firmware/V1.2)

Unless I misunderstand the implication of USB by default. As I understand it, this uploads the firmware over USB ready for flash on reboot as part of the PlatformIO build?

**Comment by thisiskeithb** [comment]
> It is wise to switch to this if the OEM manuals assume we default to non-usb? (i.e. the last section of this https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/tree/master/firmware/V1.2)

That guide is out of date since you no longer have to add in BigTreeTech's custom stepper library. It was also probably conservative about space since enabling USB on 256k chips ate up a lot of storage.

**Comment by swilkens** [comment]
> 
> 
> > It is wise to switch to this if the OEM manuals assume we default to non-usb? (i.e. the last section of this https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/tree/master/firmware/V1.2)
> 
> That guide is out of date since you no longer have to add in BigTreeTech's custom stepper library. It was also probably conservative about space since enabling USB on 256k chips ate up a lot of storage.

While that's certainly true, I also no longer use it, everybody buying this board is still being directed there.

**Comment by thisiskeithb** [comment]
BigTreeTech’s repo lags behind this one, so hopefully the guide will be updated once they pull in upstream changes.

**Comment by tpruvot** [comment]
all the shops show RCT6 chip, i dont agree forcing to 512K by default without ability to easily swap (.ld etc) its a mistake imo

**Comment by thinkyhead** [comment]
> all the shops show RCT6 chip

One point of the OP is that the chips marked "RCT6" on these boards have been found to actually have 512K. So the markings on the chip are not meaningful.

**Comment by sl1pkn07** [comment]
https://github.com/MarlinFirmware/Marlin/issues/15254

Related

Still need more info about bootloader

**Comment by thinkyhead** [comment]
So, simplest approach here is to make an extra environment, and use `-DSTM32_FLASH_SIZE=256` or `-DSTM32_FLASH_SIZE=512` to specify the flash space for each. If added to the linker flags, the `.ld` file can get this value and use `#if`/`#endif` too.

**Comment by tpruvot** [comment]
i fear read/write to upper 256Ks on RCT6 might overwrite first bytes.. like the fake SD cards....

seems very weird to me if that is true... may be fake/chinese STM32 chips else

**Comment by thinkyhead** [comment]
Please watch the linked video in which it is tested so we all are on the same page. https://www.youtube.com/watch?v=EBIahC1P2e0

**Comment by b-desconocido** [comment]
> i fear read/write to upper 256Ks on RCT6 might overwrite first bytes.. like the fake SD cards....
> 
> seems very weird to me if that is true... may be fake/chinese STM32 chips else

I'm using 128K of flash memory on legit STM32F103C8T6 for years without any issues, although datasheets specified 64K of flash memory. Many of Chinese manufacturers exploited this "feature" too, for example some smart soldering stations and transceiver modules for RC remote controllers. Original ST-Link V2 + STM32 ST-Link Utility indicated 128K of memory too, you could write whatever you want to upper memory region and read it back again. Someone could attach ST-Link V2 to SKR mini e3 v1.2, which has RCT6 chip on-board to confirm STM32 ST-Link Utility displays 512K instead of 256K of flash and post a photo + screenshot here. I'm way too lazy to disassemble my printer again

**Comment by tpruvot** [comment]
ive a SKR Mini V1.1 with a RCT6, and im sure the STLink show 256KB

**Comment by b-desconocido** [comment]
> ive a SKR Mini V1.0 with a RCT6, and im sure the STLink show 256KB

An example output for STM32F103C8T6:
image
I wish i didn't install and wire up my SKR mini e3 v1.2 yesterday, i could test anything then :( Okay, i'm disassembling my ender3pro now.

**Comment by thisiskeithb** [comment]
The linked videos/threads in my post show that despite chips self reporting 256k, you can read/write 512k without a problem.

**Comment by tpruvot** [comment]
image

**Comment by b-desconocido** [comment]
> The linked videos/threads in my post show that despite chips self reporting 256k, you can read/write 512k without a problem.

Does anyone tested the amount of RAM though? For example, C8T6 and CBT6 both have 20Kb. RET & RCT have different amount of RAM, 48KB and 64KB accordingly.

**Comment by tpruvot** [comment]
right, seems to work, built with RCT6 and changed pins to force 512KB

```
#if 1 //def MCU_STM32F103RE
  #define STM32_FLASH_SIZE (512 * 1024)
#else
  #define STM32_FLASH_SIZE (256 * 1024)
#endif
```

image

**Comment by b-desconocido** [comment]
> right, seems to work, built with RCT6 and changed pins to force 512KB
> 
> ```
> #if 1 //def MCU_STM32F103RE
>   #define STM32_FLASH_SIZE (512 * 1024)
> #else
>   #define STM32_FLASH_SIZE (256 * 1024)
> #endif
> ```
> 
> image

Could you check the RAM size, please? From address 0x20000000, size 0xC000 and 0x10000 accordingly.

**Comment by tpruvot** [comment]
image
yep, i can edit those

**Comment by tpruvot** [comment]
so maybe, the platformio entry should be renamed to RET6... if its fully compatible...

**Comment by b-desconocido** [comment]
Lol, thank you! Now i'm curious will it cause a hard fault if it is accessed directly by CPU instead

**Comment by tpruvot** [comment]
https://drive.google.com/file/d/13ReQ8lGl8vKstoi3IBOo9YVQePUI_WBr/view?usp=sharing

**Comment by b-desconocido** [comment]
If there are no issues with upper region of RAM, like faults raised on CPU access (probably HardFault exception) to these locations, it is possible to use RET linker scripts and other stuff without any problems. I'm sure ST wouldn't bother to modify their chip nor messing with obtaining yet another license from ARM. At least they didn't mess around with C8T6 MCU. It is easy to test, as, AFAIK, the stack is allocated at the topmost memory address - just change the target to RET or modify RCT .ld script to use full 64KB of RAM.

**Comment by tpruvot** [comment]
I guess its because the chips are recent... maybe some years ago some high density chips were bad and were downgraded to smaller chips...

**Comment by b-desconocido** [comment]
Or they just don't want to license yet another die because of minor differences. Or they have more flexibility to fulfill market demands quickly enough without any major changes in production process, who knows. Great bonus :) Gonna try USB composite with RET target later

**Comment by b-desconocido** [comment]
> I guess its because the chips are recent... maybe some years ago some high density chips were bad and were downgraded to smaller chips...

Can you share your bootloader, please?

**Comment by tpruvot** [comment]
i dont use any, i use stflash directly (without .py, commented in platformio.ini)

**Comment by Lord-Quake** [comment]
I used version 2019-11-10 and made this change (RC to RE) in platformio.ini:

```ini
[env:STM32F103RC_bigtree]
platform          = ststm32
framework         = arduino
board             = genericSTM32F103RE
platform_packages = tool-stm32duino
```
I deleted the ".pio" folder.
It compiled successfully.
```
DATA:    [==        ]  17.0% (used 11112 bytes from 65536 bytes)
PROGRAM: [====      ]  43.8% (used 229804 bytes from 524288 bytes)
```
My question: Would this be a safe alternative choice of modification?

**Comment by tpruvot** [comment]
you will know fast, if you use the flash eeprom... if it works you are good

File: platformio.ini [hunk 1]
```diff
@@ -293,30 +293,35 @@ upload_protocol   = serial
 
 #
 # BigTree SKR Mini V1.1 / SKR mini E3 / SKR E3 DIP (STM32F103RCT6 ARM Cortex-M3)
+# Newer boards ship with a 512K chip labeled as RCT6.
```
Comment (tpruvot) [inline platformio.ini]: not really... i got mine when its out... its a RCT6 and have 512K and 64K ram.... newer ones seems to have RET6 printed on them

Comment (thisiskeithb) [reply platformio.ini]: I didn't add that. It was just part of a header I moved.

**Comment by brew99** [comment]
I've reported on Reddit that on my SKR mini E3 V1.0 (probably one of the first production runs as I ordered it 2 days after it's release), this change doesn't open up flash to 512, in spite of the MCU being labelled RCT6.   I've tried manually editing the files that u/Steenerson has indicated (used Vanilla marlin from Nov 10) and also u/thisiskeithb PR from his repo which already had the changes.

If the firmware.bin file is less than 228kB, then all is OK, but if it goes above that, then it doesn't even load the firmware file.   I've spent the last four days trying different things, but can't get it to load anything above 228kB for a firmware file.   The .bin is replaced by the .CUR, but it didn't actual load new firmware.

Anyone have a pre-compiled .bin file for the mini E3 that I can try, that is over 228kB, using the STM32F103RC_bigtree?

Unfortunately, I don't at this time have an STlink to look at the MCU, so can't really go much further.

IMG_0011

File: platformio.ini (hunk 1 above)
Comment (steenerson) [reply platformio.ini]: I think it varies depending on the board, the latest E3 mini v1.2 boards have RCT6, just got mine 2 weeks ago. But the DIP boards are shipping with RET6 now I think

Comment (sjasonsmith) [reply platformio.ini]: That is interesting. I have both E3 DIP and E3 Mini 1.0 boards I received in early September and both have RET6 processors.

I'm guessing they have specified RCT6 for these boards, but allow RET6 as a drop-in replacement. If that is the case the part used would vary depending on which part is cheaper or more available that day.

**Comment by sl1pkn07** [comment]
@brew99 that is because is need to mod the bootloader, or if work (any can confirm?), flash the new bootloader comes with the new units

but IDK if works or need touch other things

more info https://github.com/MarlinFirmware/Marlin/pull/15398

**Comment by thisiskeithb** [comment]
@brew99: I took the configs from @steenerson's SKR Mini E3 1.2 fork and enabled a bunch of options:
```
DATA:    [===       ]  28.2% (used 13872 bytes from 49152 bytes)
PROGRAM: [=====     ]  45.6% (used 239108 bytes from 524288 bytes)
```
[skr-mini-1.2-issue-15890-brew99.zip](https://github.com/MarlinFirmware/Marlin/files/3848643/skr-mini-1.2-issue-15890-brew99.zip)

**Comment by thisiskeithb** [comment]
> more info #15398

So all SKR Mini E3 1.0 boards require modding the bootloader to load larger binaries or just for some users?

**Comment by brew99** [comment]
> @brew99: I took the configs from @steenerson's SKR Mini E3 1.2 fork and enabled a bunch of options:
> 
> ```
> DATA:    [===       ]  28.2% (used 13872 bytes from 49152 bytes)
> PROGRAM: [=====     ]  45.6% (used 239108 bytes from 524288 bytes)
> ```
> 
> [skr-mini-1.2-issue-15890-brew99.zip](https://github.com/MarlinFirmware/Marlin/files/3848643/skr-mini-1.2-issue-15890-brew99.zip)

@thisiskeithb : thanks for that.  Unfortunately, same outcome as I had found when doing my own compile.   The .bin changes to .cur, but there is no flashing red LED on the SKR indicating a loading of firmware, and the regular Marlin screen appears as in a normal turn-on.   Tried two different SD cards also

I'm guessing that I have a different bootloader?

edit: just tried my V1.0 DIP (w/ the RCT6) that has been sitting in a package (bought at same time as the mini E3), and it's the same issue.

I have not heard from any user that has the V1.0 E3 or DIP with the RCT6 that has been successful opening it up.

**Comment by sl1pkn07** [comment]
> > more info #15398
> 
> So all SKR Mini E3 1.0 boards require modding the bootloader to load larger binaries or just for some users?

users with V1.0 E3/DIP boards with RET6 mcu, like me :(

**Comment by thisiskeithb** [comment]
> users with V1.0 E3/DIP boards with RET6 mcu, like me :(

Did you mean RCT6? RET6 has more flash than RCT6.

I want to make sure I'm fully aware of all the potential issues and which mcu/board combos they apply to.

**Comment by steenerson** [comment]
@sl1pkn07 there was a [user on reddit](https://old.reddit.com/r/ender3/comments/dsmf6x/512k_flash_working_on_skr_mini_e3_v12/f6swnn5/) that reported their 1.0 DIP board w/ RE chip flashed a 244K firmware once these changes are made, I think it's not a bootloader problem unless they shipped boards with different bootloaders?

**Comment by sl1pkn07** [comment]
~~~
DATA:    [==        ]  17.6% (used 11528 bytes from 65536 bytes)
PROGRAM: [=====     ]  50.8% (used 266308 bytes from 524288 bytes)
~~~

https://github.com/sl1pkn07/Marlin/commits/sL1_Changes_SKR_E3_DIP

**Comment by sl1pkn07** [comment]
> Did you mean RCT6?

no. i say RET6 xd, see https://github.com/MarlinFirmware/Marlin/issues/15254

**Comment by thinkyhead** [comment]
For maximum flexibility, let's keep three environments: One for the RCT6 with an actual 256K flash size (in case some should appear in future), one for the mis-labeled (or mis-specified) RCT6 that has 512K, and one for the properly labeled RET6 with 512K. Maybe we can drop one of these.

Use `-DSTM32_FLASH_SIZE=512` to specify the flash size in the `build_flags` (and "`linker_flags`"?) in the environments (and we will multiply by 1024 where needed in the code). We should also be able to use `#if STM32_FLASH_SIZE == 512` in the `.ld` file.

**Comment by p3p** [comment]
> one for the mis-labeled (or mis-specified) chip that has 512K

As @sjasonsmith said it is likely these chips get binned into specific versions depending how much of it works after testing, if you do add an option to use flash outsize of an MCUs spec you should make sure the risk of random errors is clear.

**Comment by thinkyhead** [comment]
To help with that, it should not be the default. And the alternative should have "512K" in the environment name. I'm not trusting that a README in the HAL will help. But in future we can expand our build and install instructions on the website to discuss special cases.

**Comment by sjasonsmith** [comment]
> I'm not trusting that a README in the HAL will help.

I think it would only be noticed if included in a comment above the env in the platformio.ini file.

**Comment by thinkyhead** [comment]
Auto-build will also be extended to ask which variant you want to build.

**Comment by thinkyhead** [comment]
Ok, so this now presents 4 environments. The "Auto Build Marlin" plugin will ask whether you want to build for a 512K or 256K RCT6.

**Comment by tpruvot** [comment]
yep, would be too easy to do that (which works on my RCT6) : 
image

works with the stlink, unsure about sd, i dont use the skr bootloader anymore

**Comment by saeugetier** [comment]
Even though, the label says it is an RC, the silicon inside might be an RE. But it is not guaranteed to be an RE.
Logistics and testing are a big part of chip manufacturers cost. So often the manufacturer places a "bigger" silicon into a device with the labeling of a smaller part, in order to save costs. But there is no guarantee that the part has a bigger flash. So it depends on which batch you received. And as I know it from ST, there are different silicons with different sizes of flash.

So it is no fault of the Chinese manufacturer that the flash has 512kb in an RC device. It depends on the chip manufacturer. The Chinese manufacturer places the RE or RC based on what chip is currently cheaper on the market.

**Comment by saeugetier** [comment]
Placing a 512kb binary into the STM32F103RCT6 might not work in some cases, because there might be a "genuine" RC chip inside. Letting the user decide is a good idea. But you should be aware, that it might be dangerous, because the binary might be truncated.

**Comment by tpruvot** [comment]
you will know it instantly if using flash eeprom...

**Comment by swilkens** [comment]
@thisiskeithb @thinkyhead 

Any build I try directly after this merge is causing a TMC CONNECTION ERROR, including the `STM32F103RC_bigtree_NOUSB` env - which as I understand it should be the original 256K situation for this board.

The branch at this point (https://github.com/MarlinFirmware/Marlin/tree/b394a5d9afc73ca85fa259524b312e9cf91916f9) was working properly.

Oddly enough, the steppers do seem to be working.

**Comment by tatusah** [comment]
I noticed the TMC CONNECTION ERROR also, but it seems to be related to the TMCStepper being updated to 0.6.0 in Platform.io registry just recently. I noticed this issue also before this merge when testing steenerson 512K fork and TMCStepper v0.6.0. Forcing the previous version 0.5.2 corrected the issue for me. Running SKR E3 mini v1.2 board.

**Comment by swilkens** [comment]
> 
> 
> I noticed the TMC CONNECTION ERROR also, but it seems to be related to the TMCStepper being updated to 0.6.0 in Platform.io registry just recently. I noticed this issue also before this merge when testing steenerson 512K fork and TMCStepper v0.6.0. Forcing the previous version 0.5.2 corrected the issue for me. Running SKR E3 mini v1.2 board.

Looks like you are on the money, the problem went away after forcing pio to use `TMCStepper@=0.5.2`.

The steppers also do work with 0.6.0 - regardless of the errors.

Related: https://github.com/teemuatlut/TMCStepper/issues/93

**Comment by sl1pkn07** [comment]
my E3 DIP v1.0 with RET6 mcu

Screenshot_20191115_163057
IMG_20191115_163424
and flashed via stlin
~~~
Building in release mode
Checking size .pio/build/STM32F103RE/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
DATA:    [==        ]  17.6% (used 11528 bytes from 65536 bytes)
PROGRAM: [=====     ]  50.8% (used 266388 bytes from 524288 bytes)
Configuring upload protocol...
AVAILABLE: blackmagic, dfu, jlink, serial, stlink
CURRENT: upload_protocol = stlink
Uploading .pio/build/STM32F103RE/firmware.elf
xPack OpenOCD, 64-bit Open On-Chip Debugger 0.10.0+dev (2019-07-17-11:25)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
debug_level: 1

hla_swd
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x0800014c msp: 0x20001c98
** Programming Started **
** Programming Finished **
** Verify Started **
** Verified OK **
** Resetting Target **
shutdown command invoked
============================================================================================================================================== [SUCCESS] Took 23.08 seconds ==============================================================================================================================================

Environment              Status    Duration
-----------------------  --------  ------------
megaatmega2560           IGNORED
megaatmega1280           IGNORED
at90usb1286_cdc          IGNORED
at90usb1286_dfu          IGNORED
DUE                      IGNORED
DUE_USB                  IGNORED
DUE_debug                IGNORED
LPC1768                  IGNORED
LPC1769                  IGNORED
sanguino_atmega644p      IGNORED
sanguino_atmega1284p     IGNORED
melzi                    IGNORED
melzi_optiboot           IGNORED
rambo                    IGNORED
STM32F103RE              SUCCESS   00:00:23.085
~~~
working as-sis*
*EDIT: working from USB. seems power from stlink needs more power

but when put the firmware (tested upcase and lowercase), do nothing. seems the bootloader dont like the firmware wit more than 256KB

EDIT2: oh! https://github.com/DAlexis/stm32-usb-bootloader

how i can dump the original bootloader? i want to test it!

**Comment by steenerson** [comment]
@sl1pkn07 The software is also called st-link. or you can use stm32cubeprogrammer which is a prettier wrapper for st-link, either way the process is the same (assuming you're using GUI version and not command line). 

Open it up, make sure the right ST-link is selected, hit connect and it should show the device info. Leave the start address at 0x08000000, and set the size to 0x80000 to backup the full 512K firmware and bootloader simultaneously. You could set the size to 0x7000 to just grab the bootloader but might as well just dump the whole contents. Then hit connect/read and it should update the code dump on the screen based on the size you input. Then save it to a .bin file. 

Make sure the saved .bin file is the correct size and verifies correctly etc before flashing over your chip.

To restore it you'd open the .bin file, then hit 'program and verify'.

Here's the stock bootloader from a 1.2 mini e3 if you want to try that too:
https://drive.google.com/file/d/1l_Bvmi2uB8icDuGAt0mvug90LO6o5mmT/view?usp=sharing

**Comment by thinkyhead** [comment]
Alex has posted a followup video. [Check it out.](https://www.youtube.com/watch?v=q0JEx3uzgSo)

**Comment by Lord-Quake** [comment]
I worked on my printer with version `STM32F103RC_bigtree_512K` all day yesterday.
I was using it with SD and with Octoprint.
No issues concerning RAM to report.
By the way; My chip was made in China.
SKR_MINI_E3_v1 2_Chip

**Comment by LemaruX** [comment]
Anyone had any luck with the v1.0 boards yet, or tried the v1.2 bootloader on it? 

Otherwise I guess the only option is to get an ST-Link?

**Comment by saeugetier** [comment]
> Anyone had any luck with the v1.0 boards yet, or tried the v1.2 bootloader on it?

I compared the binary of the bootloader from my V1.0 with the published V1.2 bootloader binary. There is a difference of about 300 bytes. Mine V1.0 bootloader had a lock wich locks the upload of binaries greater than 228kb. But I have patched the bootloader. You have just to search for the value 0x08004000 and change it to 0x08008000. I did this task by using radare2. I cannot, whether all V1.0 have this lock. There might be more different versions of the bootloader.

Haven't tested it yet, but I assume that the V1.2 bootloader will run on a V1.0.

> Otherwise I guess the only option is to get an ST-Link?

For patching the bootloader or writing the V1.2 bootloader, you will need a JTAG/SWD debugger.

**Comment by mark078** [comment]
I have a BIGTREE_SKR_E3_DIP (v1.0) with STM32F103RET6 that blocks the firmware over 256K too.

Could you share your patched bootloader @saeugetier ? 
I looked at radare2, but I don't know how to change the address.

EDIT: the original bootloader can be found in the BIGTREETECH Git:
https://github.com/bigtreetech/BIGTREETECH-SKR-E3-DIP-V1.0/issues/2#issuecomment-515893243

So it would be really nice if you would provide the patch, or a short guide on how to do this with radare2. :)

**Comment by sl1pkn07** [comment]
> Here's the stock bootloader from a 1.2 mini e3 if you want to try that too:
> https://drive.google.com/file/d/1l_Bvmi2uB8icDuGAt0mvug90LO6o5mmT/view?usp=sharing

tested and not works for me. put my firmware into the SD, seems flashed without problem (transform to firmware.cur with the same size as aoriginal) but the firmware is never loaded (white screen)

**Comment by saeugetier** [comment]
> So it would be really nice if you would provide the patch, or a short guide on how to do this with radare2. :)

I'll just provide my patched bootloader plus my custom firmware (sorry, I haven't truncated this file):
[bootloader.zip](https://github.com/MarlinFirmware/Marlin/files/3865439/bootloader.zip)

For anybody, who want's to figure out, what I have done: I looked for places, where the values 0x08007000 (start address of application) and 0x08040000 (end address of memory) are located. Then I looked where the addresses of that places are used and disassembled the source in that locations. I saw that 0x08040000 was subtracted by 0x08007000. So I assumed that this is a length calculation. Changing the value 0x08040000 to 0x08080000 unlocked flashing up to 512kb. That was quite easy.

I don't write down, what commands I have used in radare2. You can look them up in the r2 manual.

**Comment by sl1pkn07** [comment]
> So it would be really nice if you would provide the patch, or a short guide on how to do this with radare2. :)
> 
> I'll just provide my patched bootloader plus my custom firmware (sorry, I haven't truncated this file):
> [bootloader.zip](https://github.com/MarlinFirmware/Marlin/files/3865439/bootloader.zip)

also not works for me

~~~
└───╼  openocd -f interface/stlink-v2.cfg -f target/stm32f1x.cfg -c "program bootloader.bin reset exit 0x08000000"
Open On-Chip Debugger 0.10.0
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
adapter speed: 1000 kHz
adapter_nsrst_delay: 100
none separate
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : clock speed 950 kHz
Info : STLINK v2 JTAG v21 API v2 SWIM v4 VID 0x0483 PID 0x3748
Info : using stlink api v2
Info : Target voltage: 3.274332
Info : stm32f1x.cpu: hardware has 6 breakpoints, 4 watchpoints
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x0800014c msp: 0x20000098
** Programming Started **
auto erase enabled
Info : device id = 0x10036414
Info : flash size = 512kbytes
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x2000003a msp: 0x20000098
wrote 262144 bytes from file bootloader.bin in 10.823650s (23.652 KiB/s)
** Programming Finished **
** Resetting Target **
shutdown command invoked
~~~

put my firmware [firmware.zip](https://github.com/MarlinFirmware/Marlin/files/3865586/firmware.zip) in the SD. restart the unit, sound the speaker fast, and then white screen :S

**Comment by thisiskeithb** [comment]
> also not works for me

This is why the feature is optional. It doesn't work for everyone, unfortunately.

**Comment by mark078** [comment]
> I'll just provide my patched bootloader plus my custom firmware (sorry, I haven't truncated this file)

Thanks first @saeugetier for providing the FW or bootloader. 
I will try this on my STM32F103RET6 skr e3 dip board tomorrow, then I get the stlink v2 clone.

radare2 is already hard, if you have never done anything in this direction before. :-D 
Am I on the right way?

```
$ r2 -a arm -b 16 -w bootloader-bigtreetech-e3-dip.bin -m 0x08040000
[0x08040000]> e asm.cpu = cortex
[0x08040000]> aaaa
[0x08040000]> s fcn.08040000
[0x08040000]> pd $r

```
[My ra2 Screenshots](https://www.dropbox.com/sh/dvn8edk3g6wmyzl/AABTetJLfqxyGiIGULvdVspLa?dl=0)

What surprises me is that your FW has a value "9800" and mine (bootloader-bigtreetech-e3-dip.bin) "981c", so from the right of 0x08040000 in the screenshot.

And you read out the range from (Value from 0x08080000 to 0x080842c8) and overwritten it to (0x08040000 to 0x080442c8), is it right?

I still have to figure out how to do that. But first, I test your FW... 
...and sorry for the little off topic.




@sl1pkn07 have you tried to flash the Marlin firmware via stlink without touching the bootloader?

**Comment by sl1pkn07** [comment]
Yes, zero problems (flash step) in all cases, except one bootloader pasted in the BTT issue https://github.com/bigtreetech/BIGTREETECH-SKR-E3-DIP-V1.0/issues/2#issuecomment-537288916

**Comment by brew99** [comment]
Through all of the posts here I was able to get my E3 V1.0 with RCT6 unlocked above the 224kB threshold.   With the use of the STlink V2 I was able to read a full 512kB, but the bootloader appeared to have it locked down so it wouldn't look at firmware.bin above 228kB.

With STlink I was able to find that near address 0x000012B8 the value was set to 08040000 (256kB), so I changed that to 08080000 (512kB), saved and programmed.   Took a firmware.bin at 230kb and it loaded as per normal.  Not sure if this is a valid way, but it was interesting.

IMG_0021

Also tried @steenerson V1.2 bootloader and that also worked (which I'm running now).

Thanks all

**Comment by saeugetier** [comment]
> With STlink I was able to find that near address 0x000012B8 the value was set to 08040000 (256kB), so I changed that to 08080000 (512kB), saved and programmed. Took a firmware.bin at 230kb and it loaded as per normal. Not sure if this is a valid way, but it was interesting.

Sounds legit. I changed the value at this address, too. That unlocked the bootloader to accept 512kb.

**Comment by sl1pkn07** [comment]
is strange. open the bootloader 1.2 in STM32 st-link utility (windows), in the direction 12B8, in the 8 colum, say cleary 0808000. but i'm really sure NOT works for me.

**Comment by mark078** [comment]
I received stlink today and the value under 0x000012B8 is already entered under 08080000. 
Maybe it's because I have one RET6, the other RCT6.

And @sl1pkn07 seems to have a RET6, too?

Screenshot 2019-11-21 12 30 10

**Comment by sl1pkn07** [comment]
I have RET6, yes

**Comment by mark078** [comment]
A larger (244KB) firmware over stlink flash works for me too. 

Flashing over the SD card doesn't work for me anymore.
Why I don't know. But it is not so important for me, because stlink is sufficient. 
The SD card slot still works, so that's not the reason.

Can anyone tell me how to address the second UART port in Marlin? 
So RX2 and TX2?  Would like to try [ESP3D ](https://github.com/luc-github/ESP3D) with an ESP32.

Is this enough?
```
#define SERIAL_PORT 1 
#define SERIAL_PORT_2 2 
```
https://github.com/bigtreetech/BIGTREETECH-SKR-E3-DIP-V1.0/blob/master/Hardware/SKR-E3-DIP-V1.0-SCH.pdf

**Comment by wieltar** [comment]
I have an e3 dip (ordered quite a while ago when it came out) How would I validate that the memory of it extends to 512? I have the rct6 variant of the chip so before i start flashing anything I'd like to know im not going to break anything. Preferably without ordering anything new.

I know it's from Malaysia but that's about it. 
the markings on the chip are as follows
ARM3
STMT32F103
RCT6
99222 98
MYS 99 837

**Comment by brew99** [comment]
@wieltar The easiest way is to compile a firmware.bin file that is bigger than 224kb using one of the 512K default env. in platformio file.   If the bootloader reads the firmware file, it will take the approx. 15 sec before you get the regular Marlin screen.   If the bootloader doesn't, then it will just start up as per normal.   You cannot hurt/brick it by trying firmware.

If it is one of the early runs, then it more than likely has the "locked to 256" bootoloader, and you would then need additional hardware (i.e. STlink V2) to be able to read/write to the bootloader.   I have the same early run DIP unit (and also an early run E3), and both of mine were locked down.

**Comment by Hukuma1** [comment]
I'm confused on which env to use. BTT official just updated their readme and it states to use _NOUSB variant (https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/tree/master/firmware/V1.2). Can someone please explain what the difference is? Is this if you do not plan on using the USB port the board comes with? Any penalty (besides the memory space) from enabling the USB?

**Comment by swilkens** [comment]
> 
> 
> I'm confused on which env to use. BTT official just updated their readme and it states to use _NOUSB variant (https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/tree/master/firmware/V1.2). Can someone please explain what the difference is? Is this if you do not plan on using the USB port the board comes with? Any penalty (besides the memory space) from enabling the USB?

The `_NOUSB` variant gives you only a serial port when you connect the board to the PC using a USB cable. This is the classic setup you are used to.

The other variant (without the `_NOUSB` tag) gives you a serial port AND gives you a removable storage drive of your printer's SD card when you connect to the board with a USB cable. This lets you use the serial port as well as transfer over files to the SD card over the USB cable.

The naming is a little unclear, I agree.

**Comment by Hukuma1** [comment]
Ah thank you! So all the bells and whistles (with plenty of memory to spare) would be: `STM32F103RC_bigtree_512K`

**Comment by sl1pkn07** [comment]
anyone can test this?

https://github.com/sl1pkn07/SKR_E3_DIP_Bootloader

i'm not coder,  only just edit to build in actual systems

builded with STM32CubeIDE 1.1.0

**Comment by wieltar** [comment]
I just got my stlink clone in the mail and i looked into flashing a new bootloader for my e3 dip 1.0 since it is "locked down" but it does have the 64k of ram available. 

I have no clue on how to "unlock" the bootloader and verify a flashed firmware.bin without accidentally overriding the bootloader since it's all hex mumbo to me.  Can anyone tell/guide me, how to use st link utility and flash the new firmware/bootloader?

**Pull request merged**